presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/util/TestRetryDriver.java

```diff
@@ -0,0 +1,186 @@
+        public ListenableFuture<Integer> call()
+                throws Exception
+        {
+            int attempt = getNextAttempt();
+            assertTrue(attempt < attempts.size(), "current attempt is not specified");
+            AttemptStatus attemptStatus = attempts.get(attempt);
+
+            nextAttempt();
+
+            if (attemptStatus.equals(AttemptStatus.SYNC_FAILURE)) {
+                throw new RuntimeException(SYNC_FAILURE_MESSAGE);
+            }
+
+            return executor.submit(() -> {
+                Thread.sleep(SLEEP_MILLISECONDS);
+                if (attemptStatus.equals(AttemptStatus.SUCCESS)) {
+                    return DUMMY_RESULT;
+                }
+                else {
+                    throw new RuntimeException(ASYNC_FAILURE_MESSAGE);
+                }
+            });
+        }
+    }
+
+    private boolean matchExceptionByErrorMessage(Exception exception, String errorMessage)
+    {
+        return getCausalChain(exception).stream()
+                .anyMatch(e -> e.getMessage().equals(errorMessage));
+    }
+
+    private void assertAsyncTaskExecution(int maxAttempts, List<AttemptStatus> scheduledAttempts, AttemptStatus expectedFinalStatus)
+    {
+        RetryDriver driver = RetryDriver.retry(retryExecutor)
+                .maxAttempts(maxAttempts)
+                .exponentialBackoff(minDuration, maxDuration, maxRetryTime, backOffScaleFactor);
+        MockAsyncTask task = new MockAsyncTask(scheduledAttempts, taskExecutor);
+
+        try {
+            // execute and wait on the future
+            driver.runAsync("task", task).get();
+            assertLessThanOrEqual(task.getNextAttempt(), maxAttempts, "Final attempts should be equal or less than maxAttempt. Task: " + task);
+            if (expectedFinalStatus != AttemptStatus.SUCCESS) {
+                fail("Task is expected to fail but actually succeeded. Task: " + task);
+            }
+        }
+        catch (Exception ex) {
+            assertEquals(task.getNextAttempt(), maxAttempts, "Failure must happen on max attempt. Task: " + task);
+            if (expectedFinalStatus == AttemptStatus.SUCCESS) {
+                fail("Task is expected to fail but actually succeeded. Task: " + task);
+            }
+            else if (expectedFinalStatus.equals(AttemptStatus.ASYNC_FAILURE) && !matchExceptionByErrorMessage(ex, ASYNC_FAILURE_MESSAGE)) {
+                fail("Task is expected to fail due to async error. Actual error: " + ex);
+            }
+            else if (expectedFinalStatus.equals(AttemptStatus.SYNC_FAILURE) && !matchExceptionByErrorMessage(ex, SYNC_FAILURE_MESSAGE)) {
+                fail("Task is expected to fail due to sync error. Actual error: " + ex);
+            }
+        }
+    }
+
+    @Test
+    void testRetryDriver()
+    {
+        List<AttemptStatus> twoAttempts = ImmutableList.of(AttemptStatus.ASYNC_FAILURE, AttemptStatus.SUCCESS);
+        for (int i = 1; i <= twoAttempts.size(); ++i) {
```

Comment:
It's more typical to start from 0 and use i++.

Reply:
Refactored the test.